### ISSUE TITLE
Implement Continuous Action Spaces for MFG-RL (Phase 3.3)

### ISSUE THREAD

**Issue by derrring** [derrring/MFG_PDE#61]
## 🎯 Objective

Extend MFG_PDE's reinforcement learning paradigm to support **continuous action spaces** (a ∈ ℝᵈ), bridging the gap between classical MFG theory (which assumes continuous control) and current RL implementations (which use discrete actions).

## 📊 Current Status

**Completed** ✅:
- Phase 2.1: Mean Field Q-Learning (discrete actions)
- Phase 2.2: Mean Field Actor-Critic with PPO (discrete actions)  
- Phase 2.3: Nash Q-Learning and Population PPO (discrete actions)
- Phase 3.2.1: Heterogeneous Agents (discrete actions)
- Phase 4.2.1: Traffic Flow Control (discrete actions)

**Limitation**: Current implementations optimal for |A| ≤ 20. Many real-world MFG applications require continuous control:
- Crowd navigation: velocity control a ∈ ℝ²
- Price formation: continuous price selection a ∈ ℝ₊
- Resource allocation: continuous quantities a ∈ [0, capacity]
- Autonomous vehicles: continuous steering/acceleration

## 🔬 Theoretical Foundation

Full mathematical formulation available in: `docs/theory/reinforcement_learning/continuous_action_mfg_theory.md`

**Classical MFG with Continuous Actions**:
```
-∂ₜu + H(x, ∇ₓu, m) = f(x, m)         [HJB with continuous control]
∂ₜm - Δm - ∇·(m∇ₚH(x, ∇ₓu, m)) = 0    [Fokker-Planck]

H(x, p, m) = maxₐ∈𝒜 { -p·b(x,a,m) - L(x,a,m) }
```

where 𝒜 ⊂ ℝᵈ is continuous action space.

### **Phase 3.3.1: DDPG for MFG** (3-4 weeks)
**Algorithm**: Deep Deterministic Policy Gradient + Mean Field
- Actor: μ(s, m) → a ∈ ℝᵈ (deterministic continuous policy)
- Critic: Q(s, a, m) with action as input
- Mean Field consistency: m_t = μ_t(π)
- Target networks + Ornstein-Uhlenbeck exploration

**Deliverables**:
- `mfg_pde/alg/reinforcement/algorithms/mean_field_ddpg.py`
- Theory: `docs/theory/reinforcement_learning/ddpg_mfg_formulation.md`
- Tests: `tests/unit/test_mean_field_ddpg.py`
- Example: `examples/basic/continuous_action_maze_demo.py`

### **Phase 3.3.2: TD3 for MFG** (2-3 weeks)
**Algorithm**: Twin Delayed DDPG (more stable than DDPG)
- Twin critics: Q₁(s,a,m), Q₂(s,a,m) (reduce overestimation)
- Delayed policy updates (every 2 critic updates)
- Target policy smoothing (exploration regularization)

**Deliverables**:
- `mfg_pde/alg/reinforcement/algorithms/mean_field_td3.py`
- Comparative benchmarks: DDPG vs TD3 stability analysis
- Example: Traffic flow with continuous speed control

### **Phase 3.3.3: SAC for MFG** (3-4 weeks)
**Algorithm**: Soft Actor-Critic + Mean Field (most robust)
- Maximum entropy RL: J(π) = Σᵢ E[r(sᵢ,aᵢ,mᵢ) + α H(π(·|sᵢ,mᵢ))]
- Stochastic policy: π(a|s,m) = tanh(Normal(μ(s,m), σ(s,m)))
- Automatic temperature tuning for entropy coefficient α

**Deliverables**:
- `mfg_pde/alg/reinforcement/algorithms/mean_field_sac.py`
- Entropy analysis: Effect on Nash equilibrium convergence
- Example: Crowd navigation with continuous velocity control

### **Phase 3.3.4: Multi-Population Continuous Control** (2-3 weeks)
**Extension**: Continuous actions for heterogeneous agents
- Type-specific continuous action spaces: 𝒜ᵏ ⊂ ℝᵈᵏ
- Multi-population Nash equilibrium: Best response with continuous actions
- Applications: Traffic (cars: ℝ², trucks: ℝ²), pricing (buyers: ℝ₊, sellers: ℝ₊)

**Deliverables**:
- `mfg_pde/alg/reinforcement/algorithms/multi_population_ddpg.py`
- Example: Predator-prey with continuous movement
- Example: Traffic flow with heterogeneous speed/steering control

### **Phase 3.3.5: Continuous Action Environments** (2 weeks)
**Gymnasium Environments**:
- `ContinuousActionMazeEnv`: dx/dt = a, a ∈ [-v_max, v_max]²
- `PriceFormationEnv`: Price selection a ∈ [p_min, p_max]
- `ResourceAllocationEnv`: Allocation a ∈ [0, capacity]

**Deliverables**:
- `mfg_pde/alg/reinforcement/environments/continuous_action_maze_env.py`
- Visualization: Continuous trajectory plots
- Benchmarks: Discrete vs continuous action performance

## 📈 Success Metrics

1. **Correctness**: Continuous policies satisfy Nash equilibrium (ε-Nash gap < 0.01)
2. **Performance**: Convergence within 10K episodes for standard benchmarks
3. **Stability**: TD3/SAC achieve lower variance than DDPG
4. **Scalability**: Support high-dimensional actions (d ≤ 10)
5. **Usability**: Same API design as discrete action algorithms

## 🔗 Dependencies

**Prerequisites**:
- PyTorch ≥ 1.13 (neural networks)
- Gymnasium (RL environments)
- Current MFG-RL infrastructure (mean field state tracking)

**New Requirements**:
- Continuous action exploration (Ornstein-Uhlenbeck, Gaussian noise)
- Action space clipping/squashing (tanh, sigmoid)
- Entropy computation for stochastic policies

## 📚 References

**Classical MFG Theory**:
- Lasry & Lions (2007): MFG with continuous control
- Cardaliaguet et al. (2015): Viscosity solutions for continuous action MFG
- Cardaliaguet & Lehalle (2018): Price formation MFG

**Continuous Control RL**:
- Lillicrap et al. (2015): DDPG
- Fujimoto et al. (2018): TD3  
- Haarnoja et al. (2018): SAC

**Mean Field RL**:
- Guo et al. (2019): Mean Field Multi-Agent RL
- Yang et al. (2018): Mean Field Actor-Critic

## 🎯 Timeline Estimate

**Total**: 12-16 weeks (3-4 months)

- Weeks 1-4: DDPG for MFG + theory + tests
- Weeks 5-7: TD3 for MFG + stability analysis
- Weeks 8-11: SAC for MFG + entropy analysis
- Weeks 12-14: Multi-population continuous control
- Weeks 15-16: Continuous environments + benchmarks

**Milestone**: By end of Phase 3.3, MFG_PDE will support both discrete and continuous action MFG-RL with production-quality implementations.

---

**Full Roadmap**: See `docs/theory/reinforcement_learning/continuous_action_mfg_theory.md` for detailed mathematical formulations, architecture designs, and research notes.
